---
title: "Beta多样性-限制性主坐标轴分析CPCoA"
author: "Yong-Xin Liu"
date: "2018年9月15日"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1.1 设置输入输出文件和参数

修改下面`default=`后面的文件和参数。

输入文件有标准化OTU表(distMat_norm.txt)+分组信息(design.txt)

输入文件"-i", "--input"，Input norm distMat; 标准化OTU表；

分类级别"-t", "--type"，Distance type; 距离类型, 可选bray_curtis, unifrac, unifrac_binary；

实验设计"-d", "--design"，默认为上级目录`../design.txt`，可手动修改文件位置；

分组列名"-n", "--group"，默认将design.txt中的group列作为分组信息，可修改为任意列名；

输入文件前缀"-o", "--output"，默认为空时，输出为当前目录前缀为cpcoa_bray.pdf为图片，cpcoa_bray_lable.pdf为包含样本标签的图片，方便搜索挑选异常样品。

图片宽"-w", "--width"，默认6英寸，根据图像布局可适当增大或缩小

图片高"-e", "--height"，默认4英寸，根据图像布局可适当增大或缩小

## 1.2 解析命令行

```{r parameter}
# 设置清华源加速下载
site="https://mirrors.tuna.tsinghua.edu.cn/CRAN"
# 判断命令行解析是否安装，安装并加载
if (!suppressWarnings(suppressMessages(require("optparse", character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))) {
  install.packages(p, repos=site)
  require("optparse",character.only=T) 
}
# 解析命令行
if (TRUE){
  option_list = list(
    make_option(c("-t", "--type"), type="character", default="unifrac",
                help="Distance type; 距离类型, 可选 bray_curtis, unifrac, unifrac_binary [default %default]"),   
    make_option(c("-i", "--input"), type="character", default="",
                help="Distance matrix; 距离矩阵 [default %default]"),
    make_option(c("-d", "--design"), type="character", default="../design.txt",
                help="design file; 实验设计文件 [default %default]"),
    make_option(c("-n", "--group"), type="character", default="group",
                help="name of group type; 分组列名 [default %default]"),
    make_option(c("-w", "--width"), type="numeric", default=4,
                help="Width of figure; 图片宽 [default %default]"),
    make_option(c("-e", "--height"), type="numeric", default=2.5,
                help="Height of figure; 图片高 [default %default]"),
    make_option(c("-o", "--output"), type="character", default="",
                help="output directory or prefix; 输出文件前缀, 通常会有统计表txt、矢量图pdf和位图png [default %default]")
  )
  opts = parse_args(OptionParser(option_list=option_list))
  
  # 调置如果无调设置输出，根据其它参数设置默认输出
  if (opts$input==""){opts$input=paste(opts$type,".txt", sep = "")}
  if (opts$output==""){opts$output=paste("cpcoa_",opts$type, sep = "")}
  
  # 显示输入输出确认是否正确
  print(paste("Distance type is ", opts$type,  sep = ""))
  print(paste("The distrance matrix file is ", opts$input,  sep = ""))
  print(paste("The design file is ", opts$design,  sep = ""))
  print(paste("The group name is ", opts$group,  sep = ""))
  print(paste("Output figure width ", opts$width,  sep = ""))
  print(paste("Output figure height ", opts$height,  sep = ""))
  print(paste("The output file prefix is ", opts$output, sep = ""))
}
```

# 2. 依赖关系检查、安装和加载

```{r library}
# 2.1 安装CRAN来源常用包
# 依赖包列表：参数解析、数据变换、绘图和开发包安装、安装依赖、ggplot主题
package_list = c("reshape2","ggplot2","vegan")
# 判断R包加载是否成功来决定是否安装后再加载
for(p in package_list){
  if(!suppressWarnings(suppressMessages(require(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))){
    install.packages(p, repos=site)
    suppressWarnings(suppressMessages(library(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))
  }
}

# 2.2 安装bioconductor常用包
package_list = c("ggrepel")
for(p in package_list){
  if(!suppressWarnings(suppressMessages(require(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))){
    source("https://bioconductor.org/biocLite.R")
    biocLite(p)
    suppressWarnings(suppressMessages(library(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))
  }
}

# 2.3 安装Github常用包
# 参数解析、数据变换、绘图和开发包安装
package_list = c("kassambara/ggpubr")
for(p in package_list){
  q=unlist(strsplit(p,split = "/"))[2]
  if(!suppressWarnings(suppressMessages(require(q, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))){
    install_github(p)
    suppressWarnings(suppressMessages(library(q, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))
  }
}

```



# 3. 读取输入文件

```{r read}

# 读取距离矩阵文件
distMat = read.table(opts$input, header=T, row.names= 1, sep="\t", comment.char="") 

# 读取实验设计
design = read.table(opts$design, header=T, row.names= 1, sep="\t", comment.char="") 

# 提取样品组信息,默认为group可指定
# sampFile = as.data.frame(design[,opts$group],row.names = row.names(design))
# colnames(sampFile)[1] = "group"
# sampFile$sample=row.names(design)
design$group=design[,opts$group]
sampFile=design

# 数据筛选，筛选两文件中共有
idx = rownames(sampFile) %in% colnames(distMat) # match design with alpha
sampFile = sampFile[idx,]
distMat = distMat[rownames(sampFile),rownames(sampFile)] 
```


# 4. 统计与绘图

```{r plot}
# 提取CCA中主要结果
variability_table = function(cca){
  chi = c(cca$tot.chi, cca$CCA$tot.chi, cca$CA$tot.chi)
  variability_table = cbind(chi, chi/chi[1])
  colnames(variability_table) = c("inertia", "proportion")
  rownames(variability_table) = c("total", "constrained", "unconstrained")
  return(variability_table)
}

# Constrained analysis OTU table by genotype
# 限制条件可以添加作者关注的协变量，如 +Condition(site/replicate/site+replicate)，没影响?BarcodeSequence和genotype无法计算，因为包括了group
capscale.gen = capscale(as.dist(distMat) ~ group, data=sampFile, add=F, sqrt.dist=T, distance=opts$type) 

# ANOVA-like permutation analysis
perm_anova.gen = anova.cca(capscale.gen, permutations = 1000, parallel = 4)

# generate variability tables and calculate confidence intervals for the variance
var_tbl.gen = variability_table(capscale.gen)
eig = capscale.gen$CCA$eig
variance = var_tbl.gen["constrained", "proportion"]
p.val = perm_anova.gen[1, 4]

# extract the weighted average (sample) scores
points = capscale.gen$CCA$wa[, 1:2]
points = as.data.frame(points)
colnames(points) = c("x", "y")
points = cbind(points, sampFile[match(rownames(points), rownames(sampFile)),])

# plot CPCo 1 and 2
p = ggplot(points, aes(x=x, y=y, color=group)) + geom_point() +
  labs(x=paste("CPCoA 1 (", format(100 * eig[1] / sum(eig), digits=4), "%)", sep=""),
       y=paste("CPCoA 2 (", format(100 * eig[2] / sum(eig), digits=4), "%)", sep="")) + 
  ggtitle(paste(format(100 * variance, digits=3), " % of variance; p=",format(p.val, digits=2),sep="")) + 
  theme_classic()+ 
  stat_ellipse(level=0.68)
p

# 保存pdf和png格式方便查看和编辑
ggsave(paste(opts$output, ".pdf", sep=""), p, width = opts$width, height = opts$height)
# ggsave(paste(opts$output, ".png", sep=""), p, width = opts$width, height = opts$height)
print(paste(opts$output, ".pdf finished.", sep = ""))

# 添加样品标签
p=p+geom_text_repel(label=paste(rownames(points)),colour="black",size=3)
p
# 保存pdf和png格式方便查看和编辑
ggsave(paste(opts$output, "_label.pdf", sep=""), p, width = opts$width, height = opts$height)
# ggsave(paste(opts$output, "_label.png", sep=""), p, width = opts$width, height = opts$height)
print(paste(opts$output, "_label.pdf finished.", sep = ""))


# 5. 保存图表

# 提示工作完成

print(paste0("Beta diversity: Constrianed PCoA / CAP in ",opts$type,"done!!!"))
```


If used this script, please cited:
  Zhang, J., Zhang, N., Liu, Y.X., Zhang, X., Hu, B., Qin, Y., Xu, H., Wang, H., Guo, X., Qian, J., et al. (2018).
Root microbiota shift in rice correlates with resident time in the field and developmental stage. Sci China Life Sci 61,
https://doi.org/10.1007/s11427-018-9284-4

Copyright 2016-2018 Yong-Xin Liu <metagenome@126.com>
